### PR TITLE
Bootstrap delegation for wildfly/jboss

### PR DESCRIPTION
**Pull request by realark**
* Set system property to tell jboss/wildfly to delegate to datadog bootstrap classes

### CODE REVIEW

File: dd-java-agent/instrumentation/jboss-classloading/src/main/java/datadog/trace/instrumentation/jboss/JBossClassloadingInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  public JBossClassloadingInstrumentation() {
+    super("jboss-classloading");
+  }
+
+  @Override
+  public AgentBuilder apply(final AgentBuilder agentBuilder) {
+    return agentBuilder
+        .type(named("org.jboss.modules.Module"))
+        .and(
+            new ElementMatcher<TypeDescription>() {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jboss-classloading/src/main/java/datadog/trace/instrumentation/jboss/JBossClassloadingInstrumentation.java]: My initial suggestion was to have this in a "transform" block...  Similar to what we do for injecting helper classes.

File: dd-java-agent/instrumentation/jboss-classloading/src/main/java/datadog/trace/instrumentation/jboss/JBossClassloadingInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                  if (i > 0) {
+                    ddPrefixes += ",";
+                  }
+                  ddPrefixes += BOOTSTRAP_PACKAGE_PREFIXES[i];
+                }
+                final String existing = System.getProperty("jboss.modules.system.pkgs");
+                if (null == existing) {
+                  System.setProperty("jboss.modules.system.pkgs", ddPrefixes);
+                } else if (!existing.contains(ddPrefixes)) {
+                  System.setProperty("jboss.modules.system.pkgs", existing + "," + ddPrefixes);
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jboss-classloading/src/main/java/datadog/trace/instrumentation/jboss/JBossClassloadingInstrumentation.java]: I'd suggest using a StringBuilder with the `existing` value the start of the StringBuilder.  This seems like a lot of unnecessary string concatenation.

File: dd-java-agent/instrumentation/jboss-classloading/src/main/java/datadog/trace/instrumentation/jboss/JBossClassloadingInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  };
+
+  public JBossClassloadingInstrumentation() {
+    super("jboss-classloading");
+  }
+
+  @Override
+  public AgentBuilder apply(final AgentBuilder agentBuilder) {
+    return agentBuilder
+        .type(named("org.jboss.modules.Module"))
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jboss-classloading/src/main/java/datadog/trace/instrumentation/jboss/JBossClassloadingInstrumentation.java]: Maybe a comment explaining what is significant about this class?

File: dd-java-agent/instrumentation/jboss-classloading/src/main/java/datadog/trace/instrumentation/jboss/JBossClassloadingInstrumentation.java (hunk 2 above)
Comment (realark) [reply dd-java-agent/instrumentation/jboss-classloading/src/main/java/datadog/trace/instrumentation/jboss/JBossClassloadingInstrumentation.java]: For code paths that are only hit once string concatenation isn't a big deal, but using a string builder is fine.

File: dd-java-agent/instrumentation/jboss-classloading/src/main/java/datadog/trace/instrumentation/jboss/JBossClassloadingInstrumentation.java (hunk 1 above)
Comment (realark) [reply dd-java-agent/instrumentation/jboss-classloading/src/main/java/datadog/trace/instrumentation/jboss/JBossClassloadingInstrumentation.java]: I initially did that. Problem is, bytebuddy thinks transformers which return null have failed. This was causing us to log an error.

**Comment by realark** [comment]
My comment was lost in the diff. Reposting.

I initially used a transformer to set the property. Problem is, bytebuddy thinks transformers which return null have failed. This was causing us to log an error.

It's harmless, but I think it would confuse support if they saw a failed transform error in the logs.

**Comment by realark** [comment]
It seems to have been around for a while. The version I'm testing against goes back a few years (I had to use an old version to stay java 7 compatible).

**Pull request merged**